src/key_value_store.rs

```diff
@@ -113,3 +113,148 @@ impl<
         value
     }
 }
+
+/// Changes for in-memory state.
+pub enum IndexChange<C: Eq> {
+    Add(C),
+    Remove(C),
+}
+
+/// An interface for persisting in-memory state.
+pub trait Persist<C> {
+    /// Construct a persistable state.
+    fn new() -> Self;
+    /// Insert data to the index.
+    fn insert(&mut self, change: C);
+    /// Remove data from the index.
+    fn remove(&mut self, change: C);
+    /// Revert the uncommitted changes.
+    fn revert(&mut self);
+}
+
+/// A persistable in-memory state.
+pub struct Persistable<I, C: Eq> {
+    /// In-memory index, including both committed and uncommitted changes.
+    pub index: I,
```

Comment:
I think it's slightly better design to make this field `private` and add methods on `Persistable` for querying the index. Or, implement `Deref<Target = I>`, which at least enforces that the caller can only get an immutable reference to the index and must go through the `Persist` interface to modify it.

Reply:
Makes sense. I've changed it to private and added a getter.